src/Files.App/Views/LayoutModes/ColumnViewBrowser.xaml.cs

```diff
@@ -435,20 +435,29 @@ private void ColumnViewBase_ItemTapped(object? sender, EventArgs e)
 			var column = sender as ColumnParam;
 			if (column?.ListView.FindAscendant<ColumnViewBrowser>() != this || string.IsNullOrEmpty(column.NavPathParam))
 				return;
-			
+
 			CloseUnnecessaryColumns(column);
 		}
 
 		private void CloseUnnecessaryColumns(ColumnParam column)
 		{
-			var columnPath = ((ColumnHost.ActiveBlades.Last().Content as Frame)?.Content as ColumnShellPage)?.FilesystemViewModel?.WorkingDirectory;
-			var columnFirstPath = ((ColumnHost.ActiveBlades.First().Content as Frame)?.Content as ColumnShellPage)?.FilesystemViewModel.WorkingDirectory;
-			if (string.IsNullOrEmpty(columnPath) || string.IsNullOrEmpty(columnFirstPath))
-				return;
+			var relativeIndex = -1;
+
+			if (column.Source is not null)
+			{
+				for (var i = 0; i < ColumnHost.ActiveBlades.Count && relativeIndex is -1; i++)
+				{
+					var bladeColumn = ColumnHost.ActiveBlades[i].FindDescendant<ColumnViewBase>();
+					if (bladeColumn is not null && bladeColumn == column.Source)
+						relativeIndex = i;
+				}
+			}
+
+			if (relativeIndex is -1)
+				while (relativeIndex < ColumnHost.ActiveBlades.Count &&
+					(!column.NavPathParam?.Equals(GetWorkingDirOfColumnAt(++relativeIndex)) ?? false)) ;
```

Comment:
This code is confusing and should not be handled only by the conditional clause of the while loop.
Doesn't this cause an infinite loop when `NavPathParam` is `null`?

Reply:
I'll change that but it will not cause an infinite loop because of the first condition

Reply:
Actually, I'm not sure how to make it more readable. Should I use a `do-while`?